packages/server-wallet/src/protocols/__test__/defund-channel.test.ts

```diff
@@ -0,0 +1,175 @@
+import {defaultTestConfig} from '../..';
+import {createLogger} from '../../logger';
+import {DBDefundChannelObjective, ObjectiveModel} from '../../models/objective';
+import {Store} from '../../wallet/store';
+import {testKnex as knex} from '../../../jest/knex-setup-teardown';
+import {seedAlicesSigningWallet} from '../../db/seeds/1_signing_wallet_seeds';
+import {channel} from '../../models/__test__/fixtures/channel';
+import {Channel} from '../../models/channel';
+import {MockChainService} from '../../chain-service';
+import {WalletResponse} from '../../wallet/wallet-response';
+import {ChannelDefunder} from '../defund-channel';
+import {ChallengeStatus} from '../../models/challenge-status';
+import {stateVars} from '../../wallet/__test__/fixtures/state-vars';
+import {stateWithHashSignedBy} from '../../wallet/__test__/fixtures/states';
+import {alice, bob} from '../../wallet/__test__/fixtures/signing-wallets';
+
+const logger = createLogger(defaultTestConfig());
+const timingMetrics = false;
+
+let store: Store;
+beforeEach(async () => {
+  store = new Store(
+    knex,
+    defaultTestConfig().metricsConfiguration.timingMetrics,
+    defaultTestConfig().skipEvmValidation,
+    '0'
+  );
+
+  await store.dbAdmin().truncateDB();
+  await seedAlicesSigningWallet(knex);
+});
+
+test('it should submit a conclude transaction even if there is an active challenge if there is a conclusion proof', async () => {
+  const chainService = new MockChainService();
```

Comment:
I find it really difficult to read this type of test.
- the majority of the test case is spent setting up the DB
- there are six test cases, and there appears to be a lot of repeated code
- it is difficult to look at the code, and draw a connection to the test's description

Is there no common pattern like
```
function setUpStore(testData: TestData) {
  ///
}
```

so that you can call
```
setUpStore({
  channels: [channelWithConclusionProof],
  challengeStatuses: [{channelId: c.channelId, finalizesAt: 100, challengeState}],
  ...
})
```

As something that you can do in this PR, can you at least clearly label the test setup from testing the behaviour?
```
// SETUP
const channelWithConclusionProof = ...
Store.insertChannel(channelWithConclusionProof)

// TEST
await channelDefunder.crank(obj, response);
expect(obj.status).toEqual('succeeded')
```

Reply:
I think these tests can be cleaned up! I think my preference is for descriptions with `comments`?  I feel like some attempts to DRY up test code makes it harder to understand but perhaps that a preference?

Reply:
I see your point. One particular issue here is, it is difficult to connect the test description to the test data, which is worsened because the test data is interspersed with general test setup & data insertion.

If you don't want to DRY up the test code, maybe one option is something like this?
```
test('it should submit a conclude transaction even if there is an active challenge if there is a conclusion proof', async () => {
  // Insert channel with conclusion proof
  const c = channel({
    channelNonce: 1,
    vars: [stateWithHashSignedBy([alice(), bob()])({isFinal: true})],
  });
  await Channel.query(knex).withGraphFetched('signingWallet').insert(c);

  // Insert active challenge
  const challengeState = {
    ...c.channelConstants,
    ...stateVars(),
  };
  await ChallengeStatus.insertChallengeStatus(knex, c.channelId, 100, challengeState);

  // Insert pending objective
  const obj = createPendingObjective(c.channelId);
  await knex.transaction(tx => ObjectiveModel.insert(obj, tx));

  // Test setup
  const chainService = new MockChainService();
  const channelDefunder = ChannelDefunder.create(store, chainService, logger, timingMetrics);
  const response = WalletResponse.initialize();
  const concludeSpy = jest.spyOn(chainService, 'concludeAndWithdraw');

  // Test
  await channelDefunder.crank(obj, response);

  const reloadedObjective = await store.getObjective(obj.objectiveId);
  expect(reloadedObjective.status).toEqual('succeeded');
  expect(concludeSpy).toHaveBeenCalled();
});
```

Though I'd note here that you could already DRY up the code by overwriting a `chainService` and `channelDefunder` in the `beforeEach`, since they are common to all the tests.

Reply:
I've cleaned up the tests in ec574d4. I think grouping them with descriptions and creating some simple helper functions has gone a long way to make them easier to understand

```
 PASS  src/protocols/__test__/defund-channel.test.ts (16.891 s)
  ✓ should fail the objective when the channel does not exist (555 ms)
  ✓ should fail when using non-direct funding (509 ms)
  when there is no challenge or finalized channel
    ✓ should do nothing (589 ms)
  when there is an active challenge
    ✓ should submit a conclude transaction if there is a conclusion proof (1020 ms)
    ✓ should do nothing if there is no conclusion proof (510 ms)
  when the channel is finalized on chain
    ✓ should call pushOutcome if the on-chain state is not final (494 ms)
    ✓ should do nothing if the on-chain state is final (526 ms)
```

Reply:
👍